x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/lookup/RightChunkedLeftJoin.java

```diff
@@ -0,0 +1,252 @@
+        }
+        Block[] blocks = new Block[leftHand.getBlockCount() + mergedElementCount];
+        if (rightHand.getBlockCount() != mergedElementCount + 1) {
+            throw new IllegalArgumentException(
+                "expected right hand side with [" + (mergedElementCount + 1) + "] but got [" + rightHand.getBlockCount() + "]"
+            );
+        }
+        IntVector.Builder leftFilterBuilder = null;
+        IntVector leftFilter = null;
+        IntVector.Builder insertNullsBuilder = null;
+        IntVector insertNulls = null;
+        try {
+            leftFilterBuilder = positions.blockFactory().newIntVectorBuilder(positions.getPositionCount());
+            for (int p = 0; p < positions.getPositionCount(); p++) {
+                int pos = positions.getInt(p);
+                if (pos > next) {
+                    if (insertNullsBuilder == null) {
+                        insertNullsBuilder = positions.blockFactory().newIntVectorBuilder(pos - next);
+                    }
+                    for (int missing = next; missing < pos; missing++) {
+                        leftFilterBuilder.appendInt(missing);
+                        insertNullsBuilder.appendInt(p);
+                    }
+                }
+                leftFilterBuilder.appendInt(pos);
+                next = pos + 1;
+            }
+            leftFilter = leftFilterBuilder.build();
+            int[] leftFilterArray = toArray(leftFilter);
+            insertNulls = insertNullsBuilder == null ? null : insertNullsBuilder.build();
+
+            int b = 0;
+            while (b < leftHand.getBlockCount()) {
+                blocks[b] = leftHand.getBlock(b).filter(leftFilterArray);
+                b++;
+            }
+            int rb = 1; // Skip the positions column
+            while (b < blocks.length) {
+                Block block = rightHand.getBlock(rb);
+                if (insertNulls == null) {
+                    block.mustIncRef();
+                } else {
+                    block = block.insertNulls(insertNulls);
+                }
+                blocks[b] = block;
+                b++;
+                rb++;
+            }
+            Page result = new Page(blocks);
+            blocks = null;
+            return result;
+        } finally {
+            Releasables.close(
+                blocks == null ? null : Releasables.wrap(blocks),
+                leftFilter,
+                leftFilterBuilder,
+                insertNullsBuilder,
+                insertNulls
+            );
+        }
+    }
+
+    Optional<Page> noMoreRightHandPages() {
+        if (next == leftHand.getPositionCount()) {
+            return Optional.empty();
+        }
+        BlockFactory factory = leftHand.getBlock(0).blockFactory();
+        Block[] blocks = new Block[leftHand.getBlockCount() + mergedElementCount];
+        int[] filter = IntStream.range(next, leftHand.getPositionCount()).toArray();
```

Comment:
NIT:
For consecutive Integer we could add a Block#filter(int lower, int upper) method or, depending on how easily can be JITed an Interval/Range class to have Block#filter(Interval) instead to avoid allocating the array manually.